### PR TITLE
refactor: improve performance of Drag & Drop Interactions

### PR DESCRIPTION
**Pull request by adhrinae**
### Please check if the PR fulfills these requirements
- [x] It's submitted to right branch according to our branching model
- [x] It's right issue type on title
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [ ] It does not introduce a breaking change or has description for the breaking change

### Description

- Previously, Drag & Drop(D&D) interactions rely on store updates which bring massive effects to the general performance.
- So when state change of the store occurs a lot in a short time(like D&D), the transient update can be helpful.
- Also resize, move, grid selection components should subscribe to the store directly so that it prevents unnecessary re-render starting from the parent component.

---
Thank you for your contribution to TOAST UI product. 🎉 😘 ✨

### CODE REVIEW

File: apps/calendar/src/hooks/gridSelection/gridSelection.spec.tsx [hunk 1]
```diff
@@ -96,11 +96,14 @@ describe('useGridSelection', () => {
...
 
     if (result.current) {
-      container.addEventListener('mousedown', result.current.onMouseDown);
+      container.addEventListener('mousedown', result.current);
     }
 
     return result;
   }
+  function getGridSelectionState() {
+    return store.getState().gridSelection;
```
Comment (lja1018) [inline apps/calendar/src/hooks/gridSelection/gridSelection.spec.tsx:105]: `gridSelection.timeGrid`를 바로 리턴해도 되겠네요

**Review by lja1018 — APPROVED** [review]
리뷰 완료합니다. 고생하셨습니다.

File: apps/calendar/src/components/dayGridMonth/dayGridMonth.tsx [hunk 2]
```diff
@@ -136,32 +109,31 @@ export function DayGridMonth({ options, dateMatrix = [], rowInfo = [], cellWidth
...
+              rowIndex={rowIndex}
+              cellWidthMap={cellWidthMap}
+              renderedUIModels={renderedEventUIModels}
+            />
+            <MovingEventShadow
+              dateMatrix={dateMatrix}
+              gridPositionFinder={gridPositionFinder}
+              rowIndex={rowIndex}
+              rowInfo={rowInfo}
+            />
```
[the remark this answers is not in the document]
Comment (adhrinae) [reply apps/calendar/src/components/dayGridMonth/dayGridMonth.tsx:136]: 그렇죠. 전체 Row가 리렌더 되는 것 보다 각 Row 안에 속해있는 `ResizingEventShadow` 같은애가 리렌더되는게 훨씬 비용이 저렴하니까요.

File: apps/calendar/src/components/dayGridMonth/accumulatedGridSelection.tsx [hunk 3]
```diff
@@ -8,24 +8,20 @@ import { dayGridMonthSelectionHelper } from '@src/helpers/gridSelection';
...
-
-  const gridSelectionDataByRow = gridSelections.map((gridSelection) =>
-    dayGridMonthSelectionHelper.calculateSelection(gridSelection, currentIndex, weekLength)
+export function AccumulatedGridSelection({ rowIndex, weekDates, narrowWeekend }: Props) {
+  const gridSelectionDataByRow = useStore(
+    useCallback(
+      (state) =>
+        state.gridSelection.accumulated.dayGridMonth.map((gridSelection) =>
+          dayGridMonthSelectionHelper.calculateSelection(gridSelection, rowIndex, weekDates.length)
+        ),
```
[the remark this answers is not in the document]
Comment (adhrinae) [reply apps/calendar/src/components/dayGridMonth/accumulatedGridSelection.tsx:22]: 저도 간과하다가 계속 손보고 있어요.

File: apps/calendar/src/hooks/gridSelection/gridSelection.spec.tsx (hunk 1 above)
Comment (adhrinae) [reply apps/calendar/src/hooks/gridSelection/gridSelection.spec.tsx:105]: 재언씨꺼 머지할 때는 별도의 상태를 썼던거같은데 지금은 `timeGrid` 만 꺼내쓰고 있긴 하네요.

**Pull request merged**